File: templates/problems/05_binary_search/05_search_rotated/search_rotated_test.go
package search_rotated

import "testing"

func TestSearchRotated(t *testing.T) {
	tests := []struct {
		name   string
		nums   []int
		target int
		want   int
	}{
		{"found in left half", []int{4, 5, 6, 7, 0, 1, 2}, 0, 4},
		{"found in right half", []int{4, 5, 6, 7, 0, 1, 2}, 5, 1},
		{"not found", []int{4, 5, 6, 7, 0, 1, 2}, 3, -1},
		{"single element found", []int{1}, 1, 0},
		{"single element not found", []int{1}, 0, -1},
		{"not rotated found", []int{1, 2, 3, 4, 5}, 3, 2},
		{"two elements", []int{3, 1}, 1, 1},
		{"target at pivot", []int{6, 7, 1, 2, 3, 4, 5}, 1, 2},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := searchRotated(tt.nums, tt.target); got != tt.want {
				t.Errorf("searchRotated(%v, %d) = %d, want %d", tt.nums, tt.target, got, tt.want)
			}
		})
	}
}

File: templates/problems/05_binary_search/05_search_rotated/typescript/search_rotated.test.ts
import { describe, it, expect } from "vitest";
import { searchRotated } from "./search_rotated";

describe("searchRotated", () => {
  it("found in left half", () => {
    expect(searchRotated([4, 5, 6, 7, 0, 1, 2], 0)).toBe(4);
  });

  it("found in right half", () => {
    expect(searchRotated([4, 5, 6, 7, 0, 1, 2], 5)).toBe(1);
  });

  it("not found", () => {
    expect(searchRotated([4, 5, 6, 7, 0, 1, 2], 3)).toBe(-1);
  });

  it("single element found", () => {
    expect(searchRotated([1], 1)).toBe(0);
  });

  it("single element not found", () => {
    expect(searchRotated([1], 0)).toBe(-1);
  });

  it("not rotated found", () => {
    expect(searchRotated([1, 2, 3, 4, 5], 3)).toBe(2);
  });

  it("two elements", () => {
    expect(searchRotated([3, 1], 1)).toBe(1);
  });

  it("target at pivot", () => {
    expect(searchRotated([6, 7, 1, 2, 3, 4, 5], 1)).toBe(2);
  });
});
